corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,372 @@
+import hashlib
+from abc import abstractmethod, ABCMeta, abstractproperty
+from datetime import datetime
+
+import six
+from couchdbkit import ResourceNotFound
+
+from corehq.util.couch_helpers import PaginateViewLogHandler, paginate_view
+
+
+class ResumableDocsByTypeIterator(object):
+    """Perform one-time resumable iteration over documents by type
+
+    Iteration can be efficiently stopped and resumed. The iteration may
+    omit documents that are added after the iteration begins or resumes
+    and may include deleted documents.
+
+    Example usage
+    > iterator = ResumableDocsByTypeIterator(db, doc_types, 'my-key')
+    > for doc in iterator:
+        success = process_doc(doc)
+        if success:
+            iterator.mark_doc_processed(doc)
+        else:
+            iterator.retry(doc)
+    > print iterator.progress_info
+    > iterator.discard_state()
+
+    :param db: Couchdb database.
+    :param doc_types: A list of doc type names to iterate on.
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `doc_types` to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    :param maintain_retry_state: True if the retry state of documents should be maintained
```

Comment:
Thinking about this more, I think I was misunderstanding what this was doing. It seems to be making a new way to stop the iteration before all docs have been processed (retried). This is an extension to the conventional semantics of the iterator protocol.

The normal way to stop an iteration that can later be resumed is to refrain from getting the `next` item from the iterator before the end of the iteration has been signaled by `StopIteration`. In a `for` loop this can be done with the `break` keyword or by raising an exception that breaks the loop. The iteration is complete (meaning there are no more items to be yielded) when all items have been consumed from the iterator, causing it to raise `StopIteration`. Trying to resume a completed iteration yields no items, it simply raises `StopIteration`.

`maintain_retry_state` seems to be creating a special way to allow the iterator to signal incomplete completion, meaning it raises `StopIteration` but there are still items to be retried. This complicates the exit conditions for the iteration. I think this could be trivially implemented outside of `ResumableDocsByTypeIterator`. It is even possible to maintain the extra retry state with the iterator state using the `progress_info` property.